src/include/homestore/checkpoint/cp_mgr.hpp

```diff
@@ -134,10 +112,42 @@ struct cp_mgr_super_block {
 };
 #pragma pack()
 
+class CPGuard {
+private:
+    CP* m_cp{nullptr};
+    bool m_pushed{false};
+
+    // Why we need this thread_local variable and that too of type stack?
+    // thread_local variable is needed because we wanted to make cp critical section re-entrant. So when a thread enters
+    // into a critical section and crosses methods and other code within the stack needs to enter to critical section,
+    // having this facility make sure that it uses already entered critical section within the stack.
+    //
+    // Why do we need a stack instead of only one CP* to track current critical section?
+    // It is because CPGuard can be moved from one thread to other. The thread which it is moved to can be accessed
```

Comment:
here in this example, cp1 and cp2 is the same cp or different cp?

Reply:
cp1 and cp2 are different cps.